proposals/interface-types/working-notes/scenarios/chooser.md

```diff
@@ -0,0 +1,680 @@
+
+This results in _two_ memory allocations for the shared resource: one for the
+resource itself and one for the pointer structure -- which contains a reference
+count and a raw pointer to the resource.
+
+
+```wasm
+(@interface func $stralloc-x
+  (param $str string)
+  (result owned<i32>)        ;; return an owned shared ptr of the string
+  local.get $str
+  string.size
+  call-export "malloc-x"     ;; create string entity of right size
+  own (i32)
+    call-export "free-x"     ;; temporary, until safely re-owned
+  end
+  let (local $ptr owned<i32>) (result owned<i32>)
+    local.get $str
+    local.get $ptr
+	owned.access             ;; access the owned pointer
+	local.get $str
+	string.size
+    utf8.from.string
+	local.get $ptr
+  end
+)
+
+(@interface func $shared-x
+  (param $ptr owned<i32>)
+  (result owned<i32>)        ;; return a shared ptr of the entity
+
+  local.get $ptr
+  owned.release              ;; access pointer and remove ownership
+  i32.const 1                ;; initial reference count of 1
+  call-export "shared-builder-x" ;; make it a shared ptr pair
+  own (i32)                  ;; re-own it, with a different destructor
+    call-export "shared-release-x"
+  end
+end
+)
+
+(@interface func (export "chooser")
+  (param $left string)
+  (param $right string)
+  (result string)
+
+  local.get $left
+  invoke-func $stralloc-x
+  invoke-func $shared-x     ;; make it a shared ptr
+  let (local $lpo owned<i32>) (result string)
+    local.get $right
+    invoke-func $stralloc-x
+	invoke-func $shared-x   ;; make it a shared ptr
+    let (local $rp owned<i32>) (result string)
+      ;; set up the call to the core chooser 
+      local.get $lp
+	  owned.access
+      local.get $rp
+	  owned.access
+      call-export "nondeterministic_choice"  ;; call the chooser itself
+      own (i32)               ;; own the result
+        call-export "shared-release-x"    ;; will eventually call to free string
+      end		
+      owned.access
+	  call-export "shared-get-x"
+	  call-export "access-utf8-x"
+      string.from.utf8
+    end
+  end
+)
+```
+
+For clarity, we separate out two helper functions -- `shared-x` and `stralloc-x`
+-- whose role is to create a shared pointer pair and to instantiate a memory
+string from an Interface Types's `string` entity.
+
+>Note the `-x` suffix signals that these functions are part of the exporting
+>module.
+
+>Note, like the import and export adapters themselves, they will be inlined as
+>part of the adapter fusion process.
+
+The returned value from `nondeterministic_choice` is wrapped up as an `own`ed
+allocation and lifted to a `string`.
+
+>Note we do not need to `own` the string memory of the arguments because we have
+>asserted that both the arguments to `nondeterministic_choice` will be the
+>_last_ references to the strings. However, only one of the C++ strings will be
+>deallocated -- the other is returned to us. We _do_ need to `own` the return
+>result, however.
```

Comment:
I think the ownership of the arguments and results can/should be considered orthogonally: the code as written works equally well if `nondeterministic_choice` returned a `shared_ptr<std::string>(new std::string("foo"))`, instead of one of its parameters.  In particular, the clause "because we have asserted that both the arguments to `nondeterministic_choice` will be the last references to the strings" should not be necessary: an equally-valid ABI is that the calling adapter code performs `owned.release` on arguments right before calling `nondeterministic_choice`, thereby handing off responsibility to decrement to the caller instead of having the caller *independently* increment the given pointer.  This would be slightly more efficient (less increments/decrements) and more symmetric to how returns work, so could we do that instead?

Reply:
dropped some of the text here.